src/main/java/com/avispl/symphony/dal/avdevices/power/gude/GudePDU8045Communicator.java

```diff
@@ -0,0 +1,901 @@
+		super.internalInit();
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	protected void internalDestroy() {
+		if (localExtendedStatistics != null && localExtendedStatistics.getStatistics() != null) {
+			localExtendedStatistics.getStatistics().clear();
+		}
+		if (localExtendedStatistics != null && localExtendedStatistics.getControllableProperties() != null) {
+			localExtendedStatistics.getControllableProperties().clear();
+		}
+		super.internalDestroy();
+	}
+
+
+	/**
+	 * If addressed too frequently, Sembient API may respond with 429 code, meaning that the call rate per second was reached.
+	 * Normally it would rarely happen due to the request rate limit, but when it does happen - adapter must retry the
+	 * attempts of retrieving needed information. This method retries up to 10 times with 500ms timeout in between to avoid the timeout exception
+	 *
+	 * @param url to retrieve data from
+	 * @return An instance of input class
+	 */
+	private <T> T doGetWithRetryOnUnauthorized(String url, Class<T> clazz, boolean retryOnUnauthorized) throws Exception {
+		try {
+			return doGet(url, clazz);
+		} catch (FailedLoginException e) {
+			if (retryOnUnauthorized) {
+				login();
+				return doGetWithRetryOnUnauthorized(url, clazz, false);
+			} else {
+				logger.error("retry failed");
+			}
+		}
+		return null;
+	}
+
+	/**
+	 * This method is used to log in to the camera including the Basic and Digest authentication methods
+	 */
+	private void login() {
+		try {
+			HttpClient httpClient = this.obtainHttpClient(true);
+			HttpGet httpGet = new HttpGet(buildDeviceFullPath(DeviceURL.FIRST_LOGIN));
+			HttpResponse response = null;
+
+			try {
+				response = httpClient.execute(httpGet);
+			} finally {
+				if (response instanceof CloseableHttpResponse) {
+					((CloseableHttpResponse) response).close();
+				}
+			}
+
+			Header header = response.getFirstHeader(DeviceConstant.WWW_AUTHENTICATE);
+			if (header != null) {
+				String headerResponseString = response.getFirstHeader(DeviceConstant.WWW_AUTHENTICATE).toString();
+				if (response.getStatusLine().getStatusCode() == HttpStatus.UNAUTHORIZED.value() && StringUtils.isNotNullOrEmpty(headerResponseString)) {
+					AuthorizationChallengeHandler authorizationChallengeHandler = new AuthorizationChallengeHandler(getLogin(), getPassword());
+					List<Map<String, String>> challenges = new ArrayList<>();
+					Map<String, String> challenge = authorizationChallengeHandler.parseAuthenticationOrAuthorizationHeader(headerResponseString);
+					challenges.add(challenge);
+
+					if (headerResponseString.contains(DeviceConstant.BASIC)) {
+						authorizationHeader = authorizationChallengeHandler.handleBasic();
+					}
+				}
+			}
+		} catch (HttpHostConnectException e) {
+			throw new ResourceNotReachableException(String.format("Error while connecting to %s: %s", host, e.getMessage()), e);
+		} catch (Exception e) {
+			throw new ResourceNotReachableException("Login failed" + e.getMessage(), e);
```

Comment:
this method just sends a request to get the Basic authentication header. another request will take this header for authorization, if failed login occurs it will be thrown in doGetWithRetryOnUnauthorized(). 
When an error occurs here it's not a failed login exception. Maybe a timeout exception, ect.